### PR TITLE
Fix depth sampling for ambient occlusion

### PR DESCRIPTION
**Pull request by jjhembd**
# Description

This PR fixes sampling of the depth texture in `AmbientOcclusionGenerate.glsl`.

Our ambient occlusion shader was developed when depth buffers still used linear sampling. After https://github.com/CesiumGS/cesium/pull/9966, the depth buffer uses nearest sampling. And in a WebGL2 context, linear sampling of a depth buffer is not allowed. This meant that:

- The depth value read from `depthTexture` actually corresponded to the center of the pixel.
- X and Y values were still assumed to be at un-centered positions.

The inconsistency between depth and XY values produced enough geometry error to create a background occlusion factor everywhere, including on flat surfaces which should have no occlusion.

This PR adjusts the shader calculations to explicitly shift all sampling points to pixel centers.

Along the way, I rewrote the shader to better reflect the geometry of the various inputs.

## Rendering results
Before this PR:
image

After this PR, with a WebGL2 context (default):
image

After this PR, with a WebGL1 context:
image

## Issue number and link

Partially addresses https://github.com/CesiumGS/cesium/issues/10106.

## Testing plan

- Load the [Ambient Occlusion Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html?src=Ambient%20Occlusion.html) and compare to the main branch.
- Load the [same Sandcastle with a WebGL1 context](http://localhost:8080/Apps/Sandcastle/index.html#c=vVZtb9s2EP4rnDGgDuDQerP8UjdYk/ZDi24JkGwDBn8oJTE2EZrUSMqu1+a/70i9WHaUZSuSBQYiPuS9PHfHO6ZSaIM2jG6pQm+QoFt0QTUr1vg3h/UXvdStL6QwhAmqFr0B+roQCKWA0C/mMjcMlMzQV6TonwXV5neaLLk/Q0YVFN0PFuL+5PVCLERpBadcpnc4LZSiwtywNQW7lc2PBWdEvCOG4lsl1x+0nMSe37fWFr3AC4JTb3Lq+TeeN3O/Pxa9hSiVs1vkzv1QqbqS2lwpCc7ra0OW9BNLFFE7zPTbdcLA8mWa8kKD69dFnktlaObkURUMrFMqqEVOwETJeMtEJreYcKpMdXjRu1kxjRIltxoimEmqkZAG6VIpMiuKSGkQydoiysE5lJfeYcsBrACLe0skbTLys8woh+A423olt2VIB3ZJjkhcCr6boVvCdbnPIDdCMwNgiD0HJYxAljzsuxWnYmlWFyS3kBc6TBuaX7O/6Az5tQwv1HWDengS23SCq8NhO0w4JWuqCCSt0AaCLyix1QRaXG6qlNwJSLwsDDaKpHf9hqKlXrI2UvLEiWYyLdZAEC+pec+p/TzffYAU9aozi54VO9ZM8pzvziFNTCz13sKg1mxlbqVC/dKeALchVPtoV4m2xdRgeEX05VZAMeWQ913fCp1UBxE68qAEEbJ+XyZQEhuScNr2xIk3x3SR6FSxhPaLPIO6b5Wt9RWh+6osbguR2ouGHpzrnzT3ESgdFwZ687CqcX50OTQ+FnvdVWSYCssmq3WeQ1ApEa1Q2So9Qd++dWx1VexJt5kCbrNUa90pA8VRK98ze9xEfbf+yUxzWUD3L8U6gba3V9psPqXF3q8uBRZ/Sra5jV0Kms2ntNT3t0tJvfcki9aN3+s5jnT7VKt7ddTm631PK5sEKO1oHbYFlD0EapO5Qj+YRhcEOrSG4RA6X/wgjPxgjINgGo19z48j165OR54dGD6O4ziKQjjh4DAO4W+MvcgPw5HvRaNyalQWM6Zo+qTJUw+PQt/z4iicxoHvx2HZSAEPvWA69fxw4k+8ySTy43pjPPXDKI688SSeBCM/OjCr2HJlnjI5ieLJaBp441EUxN641Aw9fDSOvJHvw140iqNxBY+88XQyDSZwNIIwHJgr8v2c3dvBqZKQJtc/Wm4NWkATHgd2e3tSj2GjdmU3Gg7RlbSPiitOYPZlcHAp0NqNNJh8G5ZBH0l26BzKkNMdut5Bha71vpEZxqmmNkBkS5hpjLp/4bubcrt8J0jxVsMC5kMA4Y5Hflnlhz1PsTUU1sY2uyzrV+pd4QJTk65QnyolVaubQpfBXC77n9/bDcQlsXOl9myGfvzqJO4/l1p6g95cmx2nZ3V7/4mt3SugULyP8RAI5hwuiB4mRXoHzqd1l0doPmyLzjO2QSx70/H6QiknWsPObcG5u3+9s/kQzj8Qrfy93FDFyc4eW/lnn0oQYzwfwrJbspmxDZO5sZ2/WVogkdmuBVhIHawtkp1Vzy3U9Jr5ENCH5+ZM5AWkfZdTS3tF07tEfgG+0FLIaQIzvYZpNnPPIUf8SBcA6nucQnZWPOLZIQLYf/K0cyj1js18J48P9XD6LtcVEUsoH7RmAla+/SJf7JcHn3ZgVGib14bwAp6CzVQcIIf86jr/DL1yBl49G8FPbvYhGH7PwNDbM9wThNfvIxybufvCHO0gRfYmv1gSH+dYPwtemOI5vIH+/wTal9dLE4OHEHrOBO4pRv+KYush9kxUYXnQ2mHd6v3tgfE3).

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

### CODE REVIEW

**Review by ggetz — CHANGES_REQUESTED** [review]
This looks great @jjhembd! I just have some small comments, but the code itself looks great!

File: packages/engine/Source/Shaders/PostProcessStages/AmbientOcclusionGenerate.glsl [hunk 1]
```diff
@@ -6,109 +8,107 @@ uniform float lengthCap;
 uniform float stepSize;
 uniform float frustumLength;
 
-in vec2 v_textureCoordinates;
```
[the remark this answers is not in the document]
Comment (jjhembd) [reply packages/engine/Source/Shaders/PostProcessStages/AmbientOcclusionGenerate.glsl:9]: See https://github.com/CesiumGS/cesium/issues/12208

**Comment by jjhembd** [comment]
@ggetz thanks for the feedback! I think I addressed it above.

**Comment by ggetz** [comment]
Awesome, thanks @jjhembd!

**Pull request merged**